contracts/Core/StakeManager.sol

```diff
@@ -103,21 +102,22 @@ contract StakeManager is Initializable, ACL, StakeStorage, StateManager, Pause {
         uint256 amount
     ) external initialized checkEpochAndState(State.Commit, epoch, parameters.epochLength()) whenNotPaused {
         require(stakers[stakerId].acceptDelegation, "Delegetion not accpected");
-        // Step 1:  Razor Token Transfer : Amount
-        require(razor.transferFrom(msg.sender, address(this), amount), "RZR token transfer failed");
+        require(stakers[stakerId].tokenAddress != address(0x0), "Staker has not staked yet");
+        emit Delegated(msg.sender, epoch, stakerId, stakers[stakerId].stake, block.timestamp);
 
-        // Step 2 : Calculate Mintable amount
+        // Step 1 : Calculate Mintable amount
         IStakedToken sToken = IStakedToken(stakers[stakerId].tokenAddress);
         uint256 totalSupply = sToken.totalSupply();
         uint256 toMint = _convertRZRtoSRZR(amount, stakers[stakerId].stake, totalSupply);
 
-        // Step 3: Increase given stakers stake by : Amount
+        // Step 2: Increase given stakers stake by : Amount
         stakers[stakerId].stake = stakers[stakerId].stake + (amount);
 
+        // Step 3:  Razor Token Transfer : Amount
+        require(razor.transferFrom(msg.sender, address(this), amount), "RZR token transfer failed");
```

Comment:
Hey @adi44 
I am not sure if this is best impl.
We had placed this transfer before
```
stakers[stakerId].stake = stakers[stakerId].stake + (amount);
```
As stake in stakeManager should be increased only when the transfer is successful
Now with this PR, we update state first and then transferFrom happens 
I am not sure what the best practice is 
But I think if its ```subtraction``` then it should be 
1. State
2. Trasnfer

If its ```addition``` then it should be 
1. Transfer
2. State

For example : Check this code 
https://github.com/maticnetwork/contracts/blob/45aa3a62264f15b6bb4d50a3fa8f6ab7b9cde829/contracts/staking/stakeManager/StakeManager.sol#L314

Here they are doing Transfer and then State as its addition

@hrishikeshio @SkandaBhat @SamAg19
 What's your opinion ?

Reply:
here they are doing addition using the internal function but we are changing in the function only.

Reply:
doing addition in internal function or doing in same function, its same.

Reply:
state changes first. interactions later.

check - effects - interactions.

doing state changes later introduces potential reentrancy vulnerability.